File: src/entities/battle.ts
import type { Army,BattleResult } from '../types/index'
import { BattleField } from '../services/battleField';
import { BattleRegistry } from '../services/battlesRegistry';

//Es necesario persistir estado interno de la batalla para ser consultados en el historial de batalla de cada ejército.
//Por eso lo creamos como una clase a ser instanciada en vez de una clase pública con métodos static.
export class Battle {
    id:string=''
    readonly attacker;
    readonly defender;
    battleResult:BattleResult = {
        winner:null,
        loser:null,
        goldAwarded:null,
        unitsLost:null
    }

    constructor(attacker:Army, defender:Army){
        
        this.attacker=attacker;
        this.defender=defender;
        
        BattleField.determineWinner(this)
        BattleField.processResult(this.battleResult)
        BattleRegistry.registryBattle(this,[attacker.getId(),defender.getId()])

    }

}

File: army_challenge/entities/battle.py
from .types import IArmy,IBattleResult
from .services.battleField import BattleField
from services.battlesRegistry import BattleRegistry

# Es necesario persistir estado interno de la batalla para ser consultados en el historial de batalla de cada ejército.
# Por eso lo creamos como una clase a ser instanciada en vez de una clase con métodos static.
class Battle:

    def __init__(attacker:IArmy, defender:IArmy)
        
        self.id: str = ''
        self.attacker = attacker
        self.defender = defender
        self.battle_result: dict[str, Optional[str]] = {
            "winner": None,
            "loser": None,
            "goldAwarded": None,
            "unitsLost": None,
        }
        
        BattleField.determineWinner(self)
        BattleField.processResult(self.battleResult)
        BattleRegistry.registryBattle(self,[attacker.id,defender.id])
